### ISSUE TITLE
Messenger depends on EventDispatcher, not PSR-14

### ISSUE THREAD

**Issue by Crell** [symfony/symfony#45963]
### Symfony version(s) affected

5.4

### Description

In the Messenger component, several classes depend on Symfony's `EventDispatcherInterface` rather than the PSR-14 dispatcher interface.  In each case the PSR interface is sufficient for the class's needs, but the current type declaration precludes using Messenger with a non-Symfony PSr-14 dispatcher.  This should be corrected to use PSR-14 directly for easier reuse.

Specifically:

* `Worker`
* `ConsumesMessagesCommand`
* `FailedMessagesRetryCommand`
* `SendMessageMiddleware`
* `SendFailedMessageForRetryListener`

### How to reproduce

See the classes in the code.

/cc @weaverryan

**Comment by weaverryan** [comment]
Yea, I don’t see a reason why we needed to use Symfony’s ED contract: PSR is all we need. I likely did this, and just wasn’t thinking.

**Comment by derrabus** [comment]
I have done this change for the Mailer component previously: #42662

Looking at what I had to do there to enable PSR-14, I'm not sure if we can/should ship this kind of change as a bugfix though. Do you need it in 5.4 or would it be enough to develop this as a feature for 6.1 or 6.2?

**Comment by derrabus** [comment]
I did some digging. The two commands you listed register event subscribers through Symfony's own mechanism.

https://github.com/symfony/symfony/blob/c9a5155f677c31202770f78921c0afd45be6625c/src/Symfony/Component/Messenger/Command/ConsumeMessagesCommand.php#L174-L197

Decoupling that is probably possible, but it's not an easy bugfix anymore.

Regarding the other classes, I agree with your assessment. They only call the dispatch method in a PSR-14 compatible manner and the `EventDispatcherInterface` appears in the constructor signature only. We can simply swap the consumed interface here.

**Comment by derrabus** [comment]
#45967

**Comment by weaverryan** [comment]
Wow! Thanks for that fast action! I know from talking with @Crell that the command class isn’t necessary to change: I believe IF he were to use Messenger, it wouldn’t include that class specifically :)

**Comment by Crell** [comment]
The Command class is low priority, I agree.  It's the others that are blockers.

As far as version, TYPO3 is currently using some 5.x components, so for as long as we're doing that we'd likely stick to the same for Messenger, should we adopt it.  (That's what I am currently researching.)  That said, 5.4 also has a lot of dependencies that really should be optional, and are in 6.x, which might push us to 6.0 for that one.  I don't know yet.

I think it's a bugfix for 6.0.x at least, and up.  It would be better to fix it in 5.4.x if possible as it gives both us and others more flexibility.  The non-Command class parts should be a trivial change as the Symfony EventDispatcher implements the PSR interface.

**Comment by derrabus** [comment]
As per https://github.com/symfony/symfony/pull/45967#issuecomment-1092442569 my change will target 6.1 now. But that should not block you if you're still on 5.4. You can install the `symfony/event-dispatcher-contracts` package which is really small and build a trivial adapter.

```php
class EventDispatcherAdapter implements \Symfony\Contracts\EventDispatcher\EventDispatcherInterface
{
    private $dispatcher;

    public function __construct(\Psr\EventDispatcher\EventDispatcherInterface $dispatcher)
    {
        $this->dispatcher = $dispatcher;
    }

    public function dispatch(object $event, string $eventName = null): object
    {
        return $this->dispatcher->dispatch($event);
    }
}
```

If my PR gets merged, you would just drop this adapter (and the `symfony/event-dispatcher-contracts` dependency) as soon as you upgrade to 6.1.
